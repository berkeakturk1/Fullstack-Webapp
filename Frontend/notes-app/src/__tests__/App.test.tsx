import React from "react";
import { render, waitFor } from "@testing-library/react";
import App from "../App";
import { MemoryRouter } from "react-router-dom";
import { act } from 'react';

// Mock fetch globally
global.fetch = jest.fn();

// Mock useParams to return a specific taskboardId
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    taskboardId: '1', // Replace with the taskboardId you expect
  }),
}));

describe("App Component Tests", () => {
  afterEach(() => {
    jest.clearAllMocks(); // Clear mocks after each test to prevent interference between tests
  });

  test("renders the App component without crashing", () => {
    render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );
  });

  test("matches the snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("fetches data from the API", async () => {
    const mockedData = [{ id: 1, title: "Test Task" }]; // Example mocked data

    (fetch as jest.Mock).mockResolvedValueOnce({
      ok: true,
      json: jest.fn().mockResolvedValueOnce(mockedData),
    });

    const { getByText } = render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith(
        "http://localhost:3001/api/workforce?taskboard_id=1"
      );
    });

    expect(getByText("Test Task")).toBeInTheDocument(); // Assuming the task title is rendered in the DOM
  });

  test("handles API errors", async () => {
    (fetch as jest.Mock).mockRejectedValueOnce(new Error("Network request failed"));

    // Spy on console.error to verify error handling
    const consoleErrorSpy = jest.spyOn(console, "error").mockImplementation(() => {});

    render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(console.error).toHaveBeenCalledWith("Error fetching workforce:", expect.any(Error));
    });

    consoleErrorSpy.mockRestore(); // Restore original console.error after the test
  });
});
