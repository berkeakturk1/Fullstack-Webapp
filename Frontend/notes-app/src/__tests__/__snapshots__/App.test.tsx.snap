// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Tests matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="app-container"
  >
    <div
      class="columns"
    >
      <div
        class="column"
        data-rbd-droppable-context-id="1"
        data-rbd-droppable-id="todo"
      >
        <div
          class="column-header"
        >
          <h2>
            TO DO 
            <span
              class="note-count"
            >
              0 ●
            </span>
          </h2>
          <button
            class="add-task-button"
          >
            +
          </button>
        </div>
      </div>
      <div
        class="column"
        data-rbd-droppable-context-id="1"
        data-rbd-droppable-id="inProgress"
      >
        <div
          class="column-header"
        >
          <h2>
            In Progress 
            <span
              class="note-count"
            >
              0 ●
            </span>
          </h2>
        </div>
      </div>
      <div
        class="column"
        data-rbd-droppable-context-id="1"
        data-rbd-droppable-id="codeReview"
      >
        <div
          class="column-header"
        >
          <h2>
            Code Review 
            <span
              class="note-count"
            >
              0 ●
            </span>
          </h2>
        </div>
      </div>
      <div
        class="column"
        data-rbd-droppable-context-id="1"
        data-rbd-droppable-id="done"
      >
        <div
          class="column-header"
        >
          <h2>
            Done 
            <span
              class="note-count"
            >
              0 ●
            </span>
          </h2>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
